### DISCUSSION TITLE
Should RuboCop allow specific MaxLineLength for individual cops?

### DISCUSSION THREAD

**Discussion by CyborgMaster** [rubocop/rubocop#10048] [Ideas]
The option for individual `MaxLineLength`s for specific cops used to be an option and was removed here:  https://github.com/rubocop-hq/rubocop/commit/9e76387952725f1a7fa54989024114549c0b7d80#diff-262d848b3b481343a51bf597d6f4d574L44.  I agree that removing it from default.yml is critical because that causes unexpected behavior, but removing the option entirely reduces RuboCop's flexibility.

In our company we set `LineLength`'s `Max` setting to 89.  The company's official guide states 80 characters, but passing RuboCop is a requirement for merging code, so we allow up to 89 to provide a little leeway and reduce resistance to following the style requirements. 

However, because the official guide for the company is 80, we don't want other cops requiring changes that would force the line length over 80.  Previously setting the `MaxLineLength` option for the `IfUnlessModifier` cop to 80 would prevent that cop flagging an offense on code that would be over 80 characters when converted to a single line modifier.  Now I have no way to handle this situation.  

Therefore I propose re-adding the following to `StatmentModifier`:

``` ruby
def max_line_length
  cop_config['MaxLineLength'] ||
    config.for_cop('Metrics/LineLength')['Max']
end
``` 

While keeping those options out of default.yml, so the default behaivor is to follow `Max` from `LineLength`.

If this is agreeable, I can provide a quick PR.

--------

## Expected behavior

No offense

## Actual behavior

```
test/test_helper.rb:363:7: C: Style/IfUnlessModifier: Favor modifier if usage when having a single-line body. Another good alternative is the usage of control flow &&/||. (https://github.com/rubocop-hq/ruby-style-guide#if-as-a-modifier)
      if Time.now - wait_start > 2
      ^^
```

## Steps to reproduce the problem

Have the following config:
``` yaml
LineLength: { Max: 89 }
Style/IfUnlessModifier: { MaxLineLength: 80 }
```
and the following code:
``` ruby
      if Time.now - wait_start > 2
        raise "Expected email to have been sent to #{address}"
      end
```

## RuboCop version

```
➜  rubocop -V
0.57.2 (using Parser 2.5.1.0, running on ruby 2.3.1 x86_64-darwin16)
```

**Comment by CyborgMaster** [comment]
Updated to include the output of `rubocop -V` instead of `rubocop -v`

**Comment by MikeMcQuaid** [comment]
Homebrew would still find this useful so just commenting so stale bot knows someone cares (although this could be combined with https://github.com/rubocop-hq/rubocop/issues/3534)

**Comment by CyborgMaster** [comment]
I would still love this as well

**Comment by CyborgMaster** [comment]
I would still like to see this implemented

**Comment by biinari** [comment]
A workaround is possible, as `IgnoredPatterns` defined in `Layout/LineLength` cop do not affect the other cops. In this configuration, `Layout/LineLength` will allow lines up to 89 characters long, but `Style/IfUnlessModifier` and other cops will treat the limit as 80 characters.

```yaml
Layout/LineLength:
  Max: 80
  IgnoredPatterns:
    - '^.{,89}$'
```

**Reply by nevans** [reply]
Thank you for this workaround!  I'll be trying it out in the near future!

**Reply by nevans** [reply]
FWIW, I updated it to allow rubocop disable comments.  This is what worked for me:

```yaml
Layout/LineLength:
  Max: 80
  AllowedPatterns:
    - '^.{,100}(?:# rubocop:(?:disable|todo) .*)?$'
```
Thanks again!

**Comment by mvz** [comment]
Since the default `Max:` for `Layout/LineLength` has been increased to 120, this would be very useful to avoid extra long lines with If/Unless modifiers.

**Comment by unquabain** [comment]
I agree with @mvz . Guard clauses (If/Unless modifiers) should be kept short, even if other lines are allowed to be long. They're inherently difficult to read, because the logic is out of order. If they get too long, they're like those trick quizzes substitute teachers give you, where the last step is to ignore all the previous steps.

**Comment by CodingAnarchy** [comment]
This feature would also be useful for cases where we want a larger overall line limit (we use 160 characters ourselves), but would not like single-line conditionals of that length.

**Comment by dvandersluis** [comment]
Is this a feature that would make sense for RuboCop?

It seems like a big deterrent for doing this would be that the config generator wouldn't know how to calculate line lengths properly if other cops could set their own line length configuration.

**Reply by mvz** [reply]
The way I imagine this working, is that none of the individual cop line lengths can be larger than the regular max line length.

**Comment by ccutrer** [comment]
I would also find such an addition useful. I would go so far as to have two separate options, controlling how auto-correct works:

MaxLineLengthForAutoCorrect: 40
MaxLineLength: 80

I.e.:
If a non-modifier if statement would convert to less than 40 characters in modifier form, auto-correct it. Otherwise, ignore it (no offense)
If a modifier if statement is currently less than 80 characters, ignore it (no offense, no auto-correct to non-modifier form).
If a modifier if statement is longer than 80 characters, convert it to non-modifier form.

This would only auto-correct things that are trivial, while leaving it up to the author when it's in a gray area. We've looked at several of our use cases, and just cause a modifier-if-usage would be under 80 characters doesn't mean it's clearer than non-modifier form. Usually this is because of method calls with arguments _and_ compound conditionals. So maybe there's a way to condition the auto-correct on those metrics?

Additionally, we currently have Layout/LineLength cop completely disabled as we're applying RuboCop to a very large codebase. Because it is disabled, Style/IfUnlessModifier is ignoring its configuration entirely, so for now we have to _also_ disable Style/IfUnlessModifier to avoid it suggesting to turn _every_ single-line-branch-body if statement into modifier form. A separate configuration option here would get around that.
